mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -26,18 +26,22 @@ import org.apache.spark.annotation.Since
 import org.apache.spark.ml.{Estimator, Model, Transformer}
 import org.apache.spark.ml.attribute.{Attribute, NominalAttribute}
 import org.apache.spark.ml.param._
-import org.apache.spark.ml.param.shared.{HasHandleInvalid, HasInputCol, HasOutputCol}
+import org.apache.spark.ml.param.shared._
 import org.apache.spark.ml.util._
-import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.{Column, DataFrame, Dataset, Encoder, Encoders, Row}
+import org.apache.spark.sql.catalyst.expressions.{If, Literal}
+import org.apache.spark.sql.expressions.Aggregator
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.types._
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.VersionUtils.majorMinorVersion
 import org.apache.spark.util.collection.OpenHashMap
 
 /**
  * Base trait for [[StringIndexer]] and [[StringIndexerModel]].
  */
 private[feature] trait StringIndexerBase extends Params with HasHandleInvalid with HasInputCol
-  with HasOutputCol {
+  with HasOutputCol with HasInputCols with HasOutputCols {
```

Comment:
Hm, do we really need HasInputCol and HasInputCols? or should HasInputCols extend HasInputCol? having one input column is a special case of having many, and this could simplify logic below.

Reply:
That's a really good point, there's the same logic in for example the `Bucketizer`. Would this make sense to do as part of this PR or a separate PR?

Reply:
OK, that can be separate. Is it necessary to actually have two different fields under the hood, one `Array[String]` and one `Array[Array[String]]`? I know the traits are going to declare them no matter what, but can the code just translate any calls that would set a single input col to modify the `Array[Array[String]]` input cols field? that could make it simpler below to manage, maybe.

Reply:
If we are talking about `inputCol` and `inputCols`, they are `String` and `Array[String]` individually. Other models that support multi-column apply an input/output pattern that `inputCol` matches `outputCol` and `inputCols` matches `outputCols`. For example, we will check if `outputCol` is set when we find `inputCol` is set. `StringIndexer` now follows other models. Making change to that like translating any calls that would set `inputCol` to modify `inputCols` will require some other changes like input/output check. I'd like to address them if it is needed, in a separate issue.

Reply:
OK, seems worth considering if we can improve handling of "HasFoo" and "HasFoos" at some point, but not critical.

Reply:
Yeah, thanks. We have few models which support multi-column. I think it is better to keep consistent behavior across them. So let's improve this part at some point if it is needed.